hoodie-common/src/main/java/com/uber/hoodie/common/util/SerializationUtils.java

```diff
@@ -16,156 +16,126 @@
+   * @param objectData the serialized object, must not be null
    * @return the deserialized object
-   * @throws IllegalArgumentException if {@code inputStream} is {@code null}
+   * @throws IllegalArgumentException if {@code objectData} is {@code null}
    * @throws HoodieSerializationException (runtime) if the serialization fails
    */
-  public static <T> T deserialize(final InputStream inputStream) {
-    if (inputStream == null) {
-      throw new IllegalArgumentException("The InputStream must not be null");
+  public static <T> T deserialize(final byte[] objectData) {
+    if (objectData == null) {
+      throw new IllegalArgumentException("The byte[] must not be null");
     }
-    ObjectInputStream in = null;
+    final KryoSerializerInstance serializerInstance = borrowKryoSerializerInstance();
     try {
-      // stream closed in the finally
-      in = new ObjectInputStream(inputStream);
-      @SuppressWarnings("unchecked") // may fail with CCE if serialised form is incorrect
-      final T obj = (T) in.readObject();
-      return obj;
-
-    } catch (final ClassCastException ex) {
-      throw new HoodieSerializationException("cannot cast class", ex);
-    } catch (final ClassNotFoundException ex) {
-      throw new HoodieSerializationException("class not found", ex);
-    } catch (final IOException ex) {
-      throw new HoodieSerializationException("unable to deserialize to object", ex);
+      return (T) serializerInstance.deserialize(objectData);
     } finally {
-      try {
-        if (in != null) {
-          in.close();
-        }
-      } catch (final IOException ex) { // NOPMD
-        // ignore close exception
-      }
+      releaseSerializerInstance(serializerInstance);
     }
   }
 
-  /**
-   * <p>
-   * Deserializes a single {@code Object} from an array of bytes.
-   * </p>
-   *
-   * <p>
-   * If the call site incorrectly types the return value, a {@link ClassCastException} is thrown from the call site.
-   * Without Generics in this declaration, the call site must type cast and can cause the same ClassCastException.
-   * Note that in both cases, the ClassCastException is in the call site, not in this method.
-   * </p>
-   *
-   * @param <T> the object type to be deserialized
-   * @param objectData the serialized object, must not be null
-   * @return the deserialized object
-   * @throws IllegalArgumentException if {@code objectData} is {@code null}
-   * @throws HoodieSerializationException (runtime) if the serialization fails
-   */
-  public static <T> T deserialize(final byte[] objectData) {
-    if (objectData == null) {
-      throw new IllegalArgumentException("The byte[] must not be null");
+  // Returns cached kryo serializer instance if available otherwise creates a new copy of it.
+  private static KryoSerializerInstance borrowKryoSerializerInstance() {
+    final KryoSerializerInstance kryoSerializerInstance = SerializationUtils.serializerRef
```

Comment:
if we make it ThreadLocal, how about just a static allocation without any atomic references/need for concurrency?

Reply:
See above comment about ThreadLocal. Let me know what you think. Will change it according to what we decide.

Reply:
commented above.

Reply:
done.